## Python Data Science Toolbox (Part 1)
&
## Python Data Science Toolbox (Part 2)

# Course Description

It's time to push forward and develop your Python chops even further. There are tons of fantastic functions in Python and its library ecosystem. However, as a data scientist, you'll constantly need to write your own functions to solve problems that are dictated by your data. You will learn the art of function writing in this first Python Data Science Toolbox course. You'll come out of this course being able to write your very own custom functions, complete with multiple parameters and multiple return values, along with default arguments and variable-length arguments. You'll gain insight into scoping in Python and be able to write lambda functions and handle errors in your function writing practice. And you'll wrap up each chapter by using your new skills to write functions that analyze Twitter DataFrames.

##  Writing your own functions
Free
7%

In this chapter, you'll learn how to write simple functions, as well as functions that accept multiple arguments and return multiple values. You'll also have the opportunity to apply these new skills to questions commonly encountered by data scientists.

    User-defined functions    50 xp
    Strings in Python    50 xp
    Recapping built-in functions    50 xp
    Write a simple function    100 xp
    Single-parameter functions    100 xp
    Functions that return single values    100 xp
    Multiple parameters and return values    50 xp
    Functions with multiple parameters    100 xp
    A brief introduction to tuples    100 xp
    Functions that return multiple values    100 xp
    Bringing it all together    50 xp
    Bringing it all together (1)    100 xp
    Bringing it all together (2)    100 xp
    Congratulations!    50 xp 
    

##  Default arguments, variable-length arguments and scope
0%

In this chapter, you'll learn to write functions with default arguments so that the user doesn't always need to specify them, and variable-length arguments so they can pass an arbitrary number of arguments on to your functions. You'll also learn about the essential concept of scope.

    Scope and user-defined functions    50 xp
    Pop quiz on understanding scope    50 xp
    The keyword global    100 xp
    Python's built-in scope    50 xp
    Nested functions    50 xp
    Nested Functions I    100 xp
    Nested Functions II    100 xp
    The keyword nonlocal and nested functions    100 xp
    Default and flexible arguments    50 xp
    Functions with one default argument    100 xp
    Functions with multiple default arguments    100 xp
    Functions with variable-length arguments (*args)    100 xp
    Functions with variable-length keyword arguments (**kwargs)    100 xp
    Bringing it all together    50 xp
    Bringing it all together (1)    100 xp
    Bringing it all together (2)    100 xp 
    

##  Lambda functions and error-handling
0%

Learn about lambda functions, which allow you to write functions quickly and on the fly. You'll also practice handling errors in your functions, which is an essential skill. Then, apply your new skills to answer data science questions.

    Lambda functions    50 xp
    Pop quiz on lambda functions    50 xp
    Writing a lambda function you already know    100 xp
    Map() and lambda functions    100 xp
    Filter() and lambda functions    100 xp
    Reduce() and lambda functions    100 xp
    Introduction to error handling    50 xp
    Pop quiz about errors    50 xp
    Error handling with try-except    100 xp
    Error handling by raising an error    100 xp
    Bringing it all together    50 xp
    Bringing it all together (1)    100 xp
    Bringing it all together (2)    100 xp
    Bringing it all together (3)    100 xp
    Bringing it all together: testing your error handling skills    50 xp
    Congratulations!    50 xp


## Course Description

In this second Python Data Science Toolbox course, you'll continue to build your Python data science skills. First, you'll learn about iterators, objects you have already encountered in the context of for loops. You'll then learn about list comprehensions, which are extremely handy tools for all data scientists working in Python. You'll end the course by working through a case study in which you'll apply all the techniques you learned in both parts of this course.

##  Using iterators in PythonLand
Free
0%

You'll learn all about iterators and iterables, which you have already worked with when writing for loops. You'll learn some handy functions that will allow you to effectively work with iterators. And you’ll finish the chapter with a use case that is pertinent to the world of data science and dealing with large amounts of data—in this case, data from Twitter that you will load in chunks using iterators.

    Introduction to iterators    50 xp
    Iterators vs Iterables    50 xp
    Iterating over iterables (1)    100 xp
    Iterating over iterables (2)    100 xp
    Iterators as function arguments    100 xp
    Playing with iterators    50 xp
    Using enumerate    100 xp
    Using zip    100 xp
    Using * and zip to 'unzip'    100 xp
    Using iterators to load large files into memory    50 xp
    Processing large amounts of Twitter data    100 xp
    Extracting information for large amounts of Twitter data    100 xp
    Congratulations!    50 xp 
    

##  List comprehensions and generators
0%

In this chapter, you'll build on your knowledge of iterators and be introduced to list comprehensions, which allow you to create complicated lists—and lists of lists—in one line of code! List comprehensions can dramatically simplify your code and make it more efficient, and will become a vital part of your Python data science toolbox. You'll then learn about generators, which are extremely helpful when working with large sequences of data that you may not want to store in memory, but instead generate on the fly.

    List comprehensions    50 xp
    Write a basic list comprehension    50 xp
    List comprehension over iterables    50 xp
    Writing list comprehensions    100 xp
    Nested list comprehensions    100 xp
    Advanced comprehensions    50 xp
    Using conditionals in comprehensions (1)    100 xp
    Using conditionals in comprehensions (2)    100 xp
    Dict comprehensions    100 xp
    Introduction to generator expressions    50 xp
    List comprehensions vs generators    50 xp
    Write your own generator expressions    100 xp
    Changing the output in generator expressions    100 xp
    Build a generator    100 xp
    Wrapping up comprehensions and generators.    50 xp
    List comprehensions for time-stamped data    100 xp
    Conditional list comprehensions for time-stamped data    100 xp 
    

##  Bringing it all together!
0%

This chapter will allow you to apply your newly acquired skills toward wrangling and extracting meaningful information from a real-world dataset—the World Bank's World Development Indicators. You'll have the chance to write your own functions and list comprehensions as you work with iterators and generators to solidify your Python data science chops.

    Welcome to the case study!    50 xp
    Dictionaries for data science    100 xp
    Writing a function to help you    100 xp
    Using a list comprehension    100 xp
    Turning this all into a DataFrame    100 xp
    Using Python generators for streaming data    50 xp
    Processing data in chunks (1)    100 xp
    Writing a generator to load data in chunks (2)    100 xp
    Writing a generator to load data in chunks (3)    100 xp
    Using pandas' read_csv iterator for streaming data    50 xp
    Writing an iterator to load data in chunks (1)    100 xp
    Writing an iterator to load data in chunks (2)    100 xp
    Writing an iterator to load data in chunks (3)    100 xp
    Writing an iterator to load data in chunks (4)    100 xp
    Writing an iterator to load data in chunks (5)    100 xp
    Final thoughts    50 xp


## In this course, the first of the Python Data Science Toolbox courses, you'll learn how to write your very own functions and you'll have the opportunity to apply these newfound skills to questions that commonly arise in Data Science contexts.  

Specifically,in this video and in the interactive exercises that follows it, you'll learn to do the following: Define a function without parameters, Define a function with single parameters, and Define a function that return a single value.  

In the next section, you'll learn how to pass multiple arguments to functions, as well as return multiple values from them.  

Lets check out Python's build-in function str(), which accepts an object such as number and returns a string object.  

As a data scientist, you'll need functions that have functionality specific to your needs.  (like I defined ff_sum() function in Efficient Python course last question.)  


We'll now see how to define functions via an example, a function that squares a number.  
  To define a function, we begin with the keyword def, followed by the function name (square in this example); This is then followed by a set of parentheses and a colon.  This piece of code is called function header.  
  To complete the function definition, lets write the function body by squaring a value, say 4, and printing the output.  
  Right now, our function does not have any parameters within the parentheses, we'll add them later.  
  Now whenever this function is called, the code in the function body is run. 

In [1]:
def square():    # <- Function Header
    new_value = 4**2
    print(new_value)
    
    
square()

16


We can add a parameter to the function definition in between the parentheses, here you see that we've added a parameter value and in the new function body, the new_value variable takes the square of value, which is then printed out.  We can now square any number that we pass to the function n_square as an argument.  

# A quick word on parameters and arguments: when you define a function, you write parameters in the function header.  When you call the function, you pass arguments into the function.  


In [3]:
def n_square(value):
    new_value = value**2
    print(new_value)
    
    
n_square(12)

144


you can have your function return the new value by adding the return keyword, followed by the value to return.  Now we can assign to a variable ss the result of the function call.  


There is another essqntial aspectof writing functions in Python: docstrings.  It's used to describe what your function does, such as the computations it performs or its return values.  These descriptions serve as documentation for your function so that anyone who read your function's docstring understands what your function does, without having to trace through all the code in the function definition.  

Function docstrings are placed in the immediate line after the function header and are placed in between triple quotation marks.  

In [6]:
def n_square(value):
    '''Return the square of a value
    '''
    new_value = value**2
    return new_value
    
    
ss = n_square(12)
print(ss)

144


## Strings in Python

In the video, you learned of another standard Python datatype, strings. Recall that these represent textual data. To assign the string 'DataCamp' to a variable company, you execute:

company = 'DataCamp'

You've also learned to use the operations + and * with strings. Unlike with numeric types such as ints and floats, the + operator concatenates strings together, while the * concatenates multiple copies of a string together. In this exercise, you will use the + and * operations on strings to answer the question below. Execute the following code in the shell:

object1 = "data" + "analysis" + "visualization"
object2 = 1 * 3
object3 = "1" * 3

What are the values in object1, object2, and object3, respectively?
Instructions
50 XP
Possible Answers

    object1 contains "data + analysis + visualization", object2 contains "1*3", object3 contains 13.
    object1 contains "data+analysis+visualization", object2 contains 3, object3 contains "13".
#    object1 contains "dataanalysisvisualization", object2 contains 3, object3 contains "111".


## Recapping built-in functions

In the video, Hugo briefly examined the return behavior of the built-in functions print() and str(). Here, you will use both functions and examine their return values. A variable x has been preloaded for this exercise. Run the code below in the console. Pay close attention to the results to answer the question that follows.

    Assign str(x) to a variable y1: y1 = str(x)
    Assign print(x) to a variable y2: y2 = print(x)
    Check the types of the variables x, y1, and y2.

What are the types of x, y1, and y2?
Instructions
50 XP
Possible Answers

    They are all str types.
    x is a float, y1 is an float, and y2 is a str.
#    x is a float, y1 is a str, and y2 is a NoneType.
    They are all NoneType types.
    
In [1]:
x
Out[1]:
4.89
In [3]:
y1 = str(x)
In [4]:
y2 = print(x)
4.89
In [5]:
type(y2)
Out[5]:
NoneType

## Write a simple function

In the last video, Hugo described the basics of how to define a function. You will now write your own function!

Define a function, shout(), which simply prints out a string with three exclamation marks '!!!' at the end. The code for the square() function that we wrote earlier is found below. You can use it as a pattern to define shout().

def square():
    new_value = 4 ** 2
    return new_value

Note that the function body is indented 4 spaces already for you. Function bodies need to be indented by a consistent number of spaces and the choice of 4 is common.

This course touches on a lot of concepts you may have forgotten, so if you ever need a quick refresher, download the Python for data science Cheat Sheet and keep it handy!
Instructions
100 XP

    Complete the function header by adding the appropriate function name, shout.
    In the function body, concatenate the string, 'congratulations' with another string, '!!!'. Assign the result to shout_word.
    Print the value of shout_word.
    Call the shout function.


In [7]:
# Define the function shout
def shout():
    """Print a string with three exclamation marks"""
    # Concatenate the strings: shout_word
    shout_word = 'congratulations' + '!!!'

    # Print shout_word
    print(shout_word)

# Call shout
shout()

congratulations!!!


## Single-parameter functions

Congratulations! You have successfully defined and called your own function! That's pretty cool.

In the previous exercise, you defined and called the function shout(), which printed out a string concatenated with '!!!'. You will now update shout() by adding a parameter so that it can accept and process any string argument passed to it. 
# Also note that shout(word), the part of the header that specifies the function name and parameter(s), is known as the signature of the function. You may encounter this term in the wild!
Instructions
100 XP

    Complete the function header by adding the parameter name, word.
    Assign the result of concatenating word with '!!!' to shout_word.
    Print the value of shout_word.
    Call the shout() function, passing to it the string, 'congratulations'.


In [11]:
# Define shout with the parameter, word
# *************************************************************************************************** #
def shout(word):   # <- shout(word) is called the signature of the function
    """Print a string with three exclamation marks"""
    # Concatenate the strings: shout_word
    shout_word = word + '!!!'

    # Print shout_word
    print(shout_word)

# Call shout with the string 'congratulations'
shout('congratulations')

congratulations!!!


## Functions that return single values

You're getting very good at this! Try your hand at another modification to the shout() function so that it now returns a single value instead of printing within the function. Recall that the return keyword lets you return values from functions. Parts of the function shout(), which you wrote earlier, are shown. 
# Returning values is generally more desirable than printing them out because, as you saw earlier, a print() call assigned to a variable has type NoneType.
Instructions
100 XP

    In the function body, concatenate the string in word with '!!!' and assign to shout_word.
    Replace the print() statement with the appropriate return statement.
    Call the shout() function, passing to it the string, 'congratulations', and assigning the call to the variable, yell.
    To check if yell contains the value returned by shout(), print the value of yell.


In [13]:
# Define shout with the parameter, word
def shout(word):
    """Return a string with three exclamation marks"""
    # Concatenate the strings: shout_word
    shout_word = word + '!!!'

    # Replace print with return
    return shout_word

# Pass 'congratulations' to shout: yell
yell = shout('congratulations')

# Print yell
print(yell)

print(type(yell))

congratulations!!!
<class 'str'>


## Multiple parameters and return values



# **At this point, you already know how to define your own functions and even return values from them.  What you'll learn next is how to pass multiple arguments to functions, as well as return not just one, but multiple values from them.  

Let's tweak the square function we've been working on a little bit more.  Suppose that, instead of simply squaring a value, we'd like to raise a value to the power of another value that's also passed to the function.  We can do this by having our function accept two parameters instead of just one.  

You should also changeyour function name and docstring to reflect this new behavior.  you can call the function by passing in two arguments because the function has two parameters, as declared in the function header.  The order in which the arguments are passed correspond to the order of the parameters in the function header.  
This means that when we call raise_to_power(2, 3), when the function is executed, 2 would be assigned to value1 and 3 to value2.  Looking at the function body, this meansthat the computation value1 to the power of value2 translates to 2 to the power of 3.  This function call then returns the value 8.  



In [15]:
def raise_to_power(value1, value2):
    '''Raise value1 to the power of value2
    '''
    new_value = value1**value2
    return new_value


raise_to_power(3, 4)

81

# You can also make your function return multiple values, you can do that by constructing object known as tuples in your functions.  A tuple is like a list, in that it can contain multiple values.  There are some difference however. 

Firstly, unlike a list, a tuple is immutable, that is you cannot modify the values in a tuple once it has been constructed.  
Secondly, while lists are defined using square brackets, tuples are constructed using a set of parentheses.  


# Additionally, you can also access individual tuple elements like you do with lists.  
Recall that with lists, you can use zero-indexing to access list elements. You can do the same thing with tuple. 

In [20]:
my_tuple = (1, 2, 3, 4, 6, 8)
[*aa] = my_tuple

a,b,c,d,e,f = my_tuple

print(a)
print(d)
print(aa)


print(my_tuple[3])

1
4
[1, 2, 3, 4, 6, 8]
4


Lets now modify the behaviorof our raise function, insted of returning just the value1 raise to the power of value2, lets also return the value2 to the power of value1.  You thus need to make raise return two values insted of one.  We can now use what we know of tuples to do this.  


We first change the name of our function and the docstring to reflect the new behavior of our function. 
We then in the function body, construct a tuple consisting of values we want the function to return, and also in the function body, we return the tuple.  

Calling the function constructeddemonstrates that it does exactly what we want.  


Now its your turn to play with writing functionthat accept multiple arguments and return multiple values.   

In [21]:
def raise_bose(value1, value2):
    '''Raise value1 to the power of value2, and vice versa
    '''
    new_value1 = value1**value2
    new_value2 = value2**value1
    
    new_tuple = (new_value1, new_value2)
    
    return new_tuple


raise_bose(2, 3)

(8, 9)

## Functions with multiple parameters

Hugo discussed the use of multiple parameters in defining functions in the last lecture. You are now going to use what you've learned to modify the shout() function further. Here, you will modify shout() to accept two arguments. Parts of the function shout(), which you wrote earlier, are shown.
Instructions
100 XP

    Modify the function header such that it accepts two parameters, word1 and word2, in that order.
    Concatenate each of word1 and word2 with '!!!' and assign to shout1 and shout2, respectively.
    Concatenate shout1 and shout2 together, in that order, and assign to new_shout.
    Pass the strings 'congratulations' and 'you', in that order, to a call to shout(). Assign the return value to yell.


In [23]:
# Define shout with parameters word1 and word2
def shout(word1, word2):
    """Concatenate strings with three exclamation marks"""
    # Concatenate word1 with '!!!': shout1
    shout1 = word1 + '!!!'
    
    # Concatenate word2 with '!!!': shout2
    shout2 = word2 + '!!!'
    
    # Concatenate shout1 with shout2: new_shout
    new_shout = (shout1, shout2)

    # Return new_shout
    return new_shout

# Pass 'congratulations' and 'you' to shout(): yell
yell = shout('congratulations', 'you')

# Print yell
print(yell)

('congratulations!!!', 'you!!!')


In [27]:
# Define shout with parameters word1 and word2
def shout(word1, word2):
    """Concatenate strings with three exclamation marks"""
    # Concatenate word1 with '!!!': shout1
    shout1 = word1 + '!!!'
    
    # Concatenate word2 with '!!!': shout2
    shout2 = word2 + '!!!'
    
    # Concatenate shout1 with shout2: new_shout
    new_shout = shout1, shout2     ### cause , is the key to construct a tuple, not the paretheses

    # Return new_shout
    return new_shout

# Pass 'congratulations' and 'you' to shout(): yell
yell = shout('congratulations', 'you')

# Print yell
print(yell)

('congratulations!!!', 'you!!!')


## A brief introduction to tuples

Alongside learning about functions, you've also learned about tuples! Here, you will practice what you've learned about tuples: how to construct, unpack, and access tuple elements. Recall how Hugo unpacked the tuple even_nums in the video:

a, b, c = even_nums

A three-element tuple named nums has been preloaded for this exercise. Before completing the script, perform the following:

    Print out the value of nums in the IPython shell. Note the elements in the tuple.
    In the IPython shell, try to change the first element of nums to the value 2 by doing an assignment: nums[0] = 2. What happens?

Instructions
100 XP

    Unpack nums to the variables num1, num2, and num3.
    Construct a new tuple, even_nums composed of the same elements in nums, but with the 1st element replaced with the value, 2.


In [32]:
nums = (3, 4, 6)


num1, num2, num3 = nums



# Construct even_nums
even_nums = 2, num2, num3


print(even_nums)

(2, 4, 6)


## Functions that return multiple values

In the previous exercise, you constructed tuples, assigned tuples to variables, and unpacked tuples. Here you will return multiple values from a function using tuples. Let's now update our shout() function to return multiple values. Instead of returning just one string, we will return two strings with the string !!! concatenated to each.

# ****************************************************************************************************
# Note that the return statement return x, y has the same result as return (x, y): the former actually packs x and y into a tuple under the hood!
Instructions
100 XP

    Modify the function header such that the function name is now shout_all, and it accepts two parameters, word1 and word2, in that order.
    Concatenate the string '!!!' to each of word1 and word2 and assign to shout1 and shout2, respectively.
    Construct a tuple shout_words, composed of shout1 and shout2.
    Call shout_all() with the strings 'congratulations' and 'you' and assign the result to yell1 and yell2 (remember, shout_all() returns 2 variables!).


In [29]:
# Define shout_all with parameters word1 and word2
def shout_all(word1, word2):
    '''Return a tuple of strings'''
    # Concatenate word1 with '!!!': shout1
    shout1 = word1 + '!!!'
    
    # Concatenate word2 with '!!!': shout2
    shout2 = word2 + '!!!'
    
    # Construct a tuple with shout1 and shout2: shout_words
    shout_words = shout1, shout2

    # Return shout_words
    return shout_words

# Pass 'congratulations' and 'you' to shout_all(): yell1, yell2
yell1, yell2 = shout_all('congratulations', 'you')

# Print yell1 and yell2
print(yell1)
print(yell2)

congratulations!!!
you!!!


## Bringing it all together



**So now you're able to write functions that accept multiple parameters and return multiple values, it is time to see how these learnt skills can be valuable in a Data Science context.  In the following exercrises, you'll write a function that analyzes a DataFrame of Twitter data.  

The function that you'll write will return a dictionary containing data of how many times each language was used across all the tweets in the DataFrame.  We'll see later in this course that we can generalize such a function to count occurances of any items in a DataFrame column.  



Lets now recap the basic ingredients of a function.  
We have a function header which begins with the keyword def, this is followedby the function name, parameters in parentheses and a colon.  
We than have the function body, which contains docstrings enclosed in tiple quotation marks; docstrings describe what the function does; the rest of function body performs the computation that the function does; the function body then closes with the keyword return, followed by the value or values returned by the function.  


Happy function writing


## Bringing it all together (1)

You've got your first taste of writing your own functions in the previous exercises. You've learned how to add parameters to your own function definitions, return a value or multiple values with tuples, and how to call the functions you've defined.

In this and the following exercise, you will bring together all these concepts and apply them to a simple data science problem. You will load a dataset and develop functionalities to extract simple insights from the data.

For this exercise, your goal is to recall how to load a dataset into a DataFrame. The dataset contains Twitter data and you will iterate over entries in a column to build a dictionary in which the keys are the names of languages and the values are the number of tweets in the given language. The file tweets.csv is available in your current directory.

Be aware that this is real data from Twitter and as such there is always a risk that it may contain profanity or other offensive content (in this exercise, and any following exercises that also use real Twitter data).
Instructions
100 XP

    Import the pandas package with the alias pd.
    Import the file 'tweets.csv' using the pandas function read_csv(). Assign the resulting DataFrame to df.
    Complete the for loop by iterating over col, the 'lang' column in the DataFrame df.
    Complete the bodies of the if-else statements in the for loop: if the key is in the dictionary langs_count, add 1 to the value corresponding to this key in the dictionary, else add the key to langs_count and set the corresponding value to 1. Use the loop variable entry in your code.


In [ ]:
# Import pandas
import pandas as pd

# Import Twitter data as DataFrame: df
df = pd.read_csv('tweets.csv')

# Initialize an empty dictionary: langs_count
langs_count = {}

# Extract column from DataFrame: col
col = df['lang']

# Iterate over lang column in DataFrame
for entry in col:

    # If the language is in langs_count, add 1 
    if entry in langs_count.keys():
        langs_count[entry] += 1
    # Else add the language to langs_count, set the value to 1
    else:
        langs_count[entry] = 1

# Print the populated dictionary
print(langs_count)


# Return
# {'en': 97, 'et': 1, 'und': 2}

In [ ]:
# df.columns

# Return:
#Index(['contributors', 'coordinates', 'created_at', 'entities', 'extended_entities', 'favorite_count', 'favorited', 'filter_level', 'geo', 'id', 'id_str', 'in_reply_to_screen_name',
#       'in_reply_to_status_id', 'in_reply_to_status_id_str', 'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status', 'lang', 'place', 'possibly_sensitive', 'quoted_status',
#       'quoted_status_id', 'quoted_status_id_str', 'retweet_count', 'retweeted', 'retweeted_status', 'source', 'text', 'timestamp_ms', 'truncated', 'user'],
#      dtype='object')

## Bringing it all together (2)

Great job! You've now defined the functionality for iterating over entries in a column and building a dictionary with keys the names of languages and values the number of tweets in the given language.

In this exercise, you will define a function with the functionality you developed in the previous exercise, return the resulting dictionary from within the function, and call the function with the appropriate arguments.

For your convenience, the pandas package has been imported as pd and the 'tweets.csv' file has been imported into the tweets_df variable.
Instructions
100 XP

    Define the function count_entries(), which has two parameters. The first parameter is df for the DataFrame and the second is col_name for the column name.
    Complete the bodies of the if-else statements in the for loop: if the key is in the dictionary langs_count, add 1 to its current value, else add the key to langs_count and set its value to 1. Use the loop variable entry in your code.
    Return the langs_count dictionary from inside the count_entries() function.
    Call the count_entries() function by passing to it tweets_df and the name of the column, 'lang'. Assign the result of the call to the variable result.


In [ ]:
# Define count_entries()
# How can we take a DataFrame and a DF column name as input argument,  Think, Think, Think, 
# *************************************************************************************************** #
def count_entries(df, col_name):
    """Return a dictionary with counts of 
    occurrences as value for each key."""

    # Initialize an empty dictionary: langs_count
    langs_count = {}
    
    # Extract column from DataFrame: col
    col = df[col_name]
    
    # Iterate over lang column in DataFrame
    for entry in col:

        # If the language is in langs_count, add 1
        if entry in langs_count.keys():
            langs_count[entry] += 1
        # Else add the language to langs_count, set the value to 1
        else:
            langs_count[entry] = 1

    # Return the langs_count dictionary
    return langs_count

# Call count_entries(): result
result = count_entries(tweets_df, 'lang')

# Print the result
print(result)


# Return:
# {'en': 97, 'et': 1, 'und': 2}

## Congratulations!



Congratulations, you're now a bonafide Python function writer.  On top of that, you have just written you very first DataScience function.  At this point, although you can write basic functions, you've really touched the surface of function writing capabilities.  



# ****************************************************************************************************
# In the following chapter, you'll learn how to write functions that have default arguments so that when you call them, you dont always have to specify all the parameters, you'll learn how write functions that can accept an arbitrary number of parameters and how to nest functions within one another, on top of this, you'll learn how to handle errors when writing functions, which will make your functions as robust as they need to be.  

Moreover, you'll see the importance of such techniques in DataScience by writing functions that are pertinent to the DataScience sphere like Twitter DataFrame analysis that you just performed.  


## Scope and user-defined functions



At this point, you know how to define your own functions - but not only that, you know how to write functions with multiple parameyters and can return multiple values using tuples.  


# We'll now talk about the idea of scope in the context of user-defined functions. 
You've been defining variables in your programs and so far, you've been using these variables without any problems.  However, one of the thing you should know is that not all objects that you defined are always accessable everywhere in a program.  

**Enter the idea of scope, which tells you which part of a program an object or a name may be accessed.  
Names refer to the variables or more generally, objects such as functions that are defined in your program, for example a variable x has a name, as does the function sum.  




# There are three types of scopes that you should know.  
The first one is the idea of the global scope.  A name that is in the global scope means that it is defined in the main body of a script or a Python program.  

The second one, is the local scope.  A name that is in a local scope means that it is defined within a function.  Once the execution of a function is done, any name inside the local scope ceases to exist, which means you cannot access these names outside the function definition.  

The third is something called the built-in scope; this consists of names in the pre-defined build-ins module Python provides, such as print and sum. 


In [33]:
def square(value):
    '''Return the square of value'''
    new_value = value**2
    return new_value


square(12)

144


You'll play around with the built-ins module in the interactive exercise, lets look at a couple of examples to clarify these definitions.  Lets check out our example function square from earlier.  


# ****************************************************************************************************
# We define the function and then call it, if we then try to access the variable name new_value after function execution, the name is not accessable.  This is because it was defined only within the local scope of the function.  The name new_value was not defined globally.  

In [35]:
new_value = 118


def square(value):
    '''Return the square of value'''
    new_value = value**2   # global csope name new_value updated
    return new_value


square(12)

144

In [36]:
new_value

118

Now what if we define the name new_value globally before defining and calling the function?  In short, any time we call the name in the global scope, it will access the name in the global, such as you see here.  

Anytime we call the name in the local scope of the function, it will look first in the local scope, that' why calling square(12) results in 144 and not 118.  

If Python can not find the name in the local scope, it will then and only then looking in the global scope.  

Here for example, we access new_value defined globally with the function square.  Note that the global value accessed is the value at the time the function is called, not the value when the function is defined.  Thus if we reassign new_value and call the square function, we see that the new value of new_value is accessed.  



In [37]:
new_value = 10

def square(value):
    '''Returns the square of value'''
    new_value1 = new_value**2    # Just as an example, if global scope value was used in a function
    return new_value1


square(12)

100

In [38]:
new_value = 12

square(7)

144


# To recap, when we reference a name, first the local scope is searched, then the global, If the nameis in neither then the built-in scope is searched.  

Now what if we want to alter the value of a global name within a function call?  This is where the keyword global comes in play.  


Within the function definition, we use keyword global followed by the name of the global variable that we wish to access and alter.  

In [40]:
new_value = 10

def square(value):
    '''Return the square of a value'''
    
    global new_value
    new_value2 = new_value**2
    return new_value2


square(12)

new_value = 3

square(18)

9

## Pop quiz on understanding scope

In this exercise, you will practice what you've learned about scope in functions. The variable num has been predefined as 5, alongside the following function definitions:

def func1():
    num = 3
    print(num)

def func2():
    global num
    double_num = num * 2
    num = 6
    print(double_num)

Try calling func1() and func2() in the shell, then answer the following questions:

    What are the values printed out when you call func1() and func2()?    3 and 10
    What is the value of num in the global scope after calling func1() and func2()?    6

Instructions
50 XP
Possible Answers

    func1() prints out 3, func2() prints out 6, and the value of num in the global scope is 3.
    func1() prints out 3, func2() prints out 3, and the value of num in the global scope is 3.
    func1() prints out 3, func2() prints out 10, and the value of num in the global scope is 10.
#    func1() prints out 3, func2() prints out 10, and the value of num in the global scope is 6.
    

In [42]:
num = 5


def func1():
    num = 3
    print(num)

def func2():
    global num
    double_num = num * 2
    num = 6
    print(double_num)

    
    
func1()
func2()
num

3
10


6

In [ ]:
# Define count_entries()
def count_entries(df, col_name):
    """Return a dictionary with counts of 
    occurrences as value for each key."""

    # Initialize an empty dictionary: langs_count
    langs_count = {}
    
    # Extract column from DataFrame: col
    col = df[col_name]
    
    # Iterate over lang column in DataFrame
    for entry in col:

        # If the language is in langs_count, add 1
        if entry in langs_count.keys():
            langs_count[entry] += 1
        # Else add the language to langs_count, set the value to 1
        else:
            langs_count[entry] = 1

    # Return the langs_count dictionary
    return langs_count

# Call count_entries(): result

import time
start = time.time()
result = count_entries(tweets_df, 'lang')
t_spend = time.time() - start
print(t_spend)

import line_profiler

%load_ext line_profiler
%lprun -f count_entries count_entries(tweets_df, 'lang')

# Print the result
print(result)

## The keyword global

Let's work more on your mastery of scope. In this exercise, you will use the keyword global within a function to alter the value of a variable defined in the global scope.
Instructions
100 XP

    Use the keyword global to alter the object team in the global scope.
    Change the value of team in the global scope to the string "justice league". Assign the result to team.
    Hit the Submit button to see how executing your newly defined function change_team() changes the value of the name team!


In [43]:
# Create a string: team
team = "teen titans"

# Define change_team()
def change_team():
    """Change the value of the global variable team."""

    # Use team in global scope
    global team

    # Change the value of team in global: team
    team = 'justice league'
# Print team
print(team)

# Call change_team()
change_team()

# Print team
print(team)

teen titans
justice league


## Python's built-in scope

Here you're going to check out Python's built-in scope, which is really just a built-in module called builtins. However, to query builtins, you'll need to import builtins 'because the name builtins is not itself built in…No, I’m serious!' (Learning Python, 5th edition, Mark Lutz). After executing import builtins in the IPython Shell, execute dir(builtins) to print a list of all the names in the module builtins. Have a look and you'll see a bunch of names that you'll recognize! Which of the following names is NOT in the module builtins?
Instructions
50 XP
Possible Answers

    'sum'
    'range'
#    'array'
    'tuple'

In [45]:
import builtins

print(dir(builtins))

['ArithmeticError', 'AssertionError', 'AttributeError', 'BaseException', 'BlockingIOError', 'BrokenPipeError', 'BufferError', 'BytesWarning', 'ChildProcessError', 'ConnectionAbortedError', 'ConnectionError', 'ConnectionRefusedError', 'ConnectionResetError', 'DeprecationWarning', 'EOFError', 'Ellipsis', 'EnvironmentError', 'Exception', 'False', 'FileExistsError', 'FileNotFoundError', 'FloatingPointError', 'FutureWarning', 'GeneratorExit', 'IOError', 'ImportError', 'ImportWarning', 'IndentationError', 'IndexError', 'InterruptedError', 'IsADirectoryError', 'KeyError', 'KeyboardInterrupt', 'LookupError', 'MemoryError', 'ModuleNotFoundError', 'NameError', 'None', 'NotADirectoryError', 'NotImplemented', 'NotImplementedError', 'OSError', 'OverflowError', 'PendingDeprecationWarning', 'PermissionError', 'ProcessLookupError', 'RecursionError', 'ReferenceError', 'ResourceWarning', 'RuntimeError', 'RuntimeWarning', 'StopAsyncIteration', 'StopIteration', 'SyntaxError', 'SyntaxWarning', 'SystemError

## Nested functions
# ****************************************************************************************************


**Now that you've come to grips with scope, both local and global, its time to dive a bit deeper.  

What if we have a function inner() defined within another function outer() and we reference a name x in the inner function?  


**The answer is intuitive, Python searchesthe local scope of the function inner(), then if it doesnt find x, it searches the scope of the function outer, which is called an exclosing function, because it encloses the function inner().  If Python cant find x in the scope of the enclosing function, it only then searches the global scope and then the built-in scope.  



# Why are we even nesting functions?  There are number of good reasons to do so.  

**Lets say that we want to use a process a number of times within a function.**
For example, we want a function that takes 3 numbers as parameters and performs the same function on each of them.  One way would be write out the computation 3 times, but this definitely does not scale if you need to perform the computation many times.  What we can do instead is to define a inner function within our function definition, such as we did below, and call it when necessary.  This is called nested function.  
The syntax for the inner function is exactly the same as for the any other function.  



In [ ]:
x = 1

def outer():
    '''Outer function'''
    x = 123
    
    def inner():
        '''Inner function'''
        y = x**2
        return y

In [46]:
def mod2plus5(x1, x2, x3):
    new_x1 = x1%2 + 5
    new_x2 = x2%2 + 5
    new_x3 = x3%2 + 5
    
    return new_x1, new_x2, new_x3


mod2plus5(7, 8, 9)

(6, 5, 6)

In [47]:
def mod2plus5(x1, x2, x3):
    '''Returns the remainder plus 5 of three values'''
    
    def inner(x):
        '''Returns the remainder plus 5'''
        return x%2 +5
    
    return inner(x1), inner(x2), inner(x3)


mod2plus5(8, 9, 10)

(5, 6, 5)


# Lets now look at another important use case of nested functions. 

In this example, we define a function raise_val, which contains an inner function called inner.  Now look at what raise_val returns: it returns the inner function inner.  The raise_val takes an argument n and create a function inner that returns the nth power of any number.  
That's a bit complicated and will be clearer when we use the function raise_val.  Passing the number 2 to raise_val create a function that squares any number, similarly, passing the number 3 to raise_val creates a function that cubes any number.  



In [48]:
def raise_val(n):
    '''Return the inner func'''
    
    def inner(x):
        '''Raise x to the power of n'''
        raised = x**n
        return raised
    return inner


ff = raise_val(3)
ff(2)

8


One interesting detail: when we call the function ff, it remembers the value n=3, although the enclosing scope defined by raise_val and to which n=3 is local, has finished execution.  This is a subtlety referred to as a closure in Computer Science circles and shouldnt concern you too much.  It is worth mentioning, however, as you may encounter it out there in the wild.  

Recall from our discussion of scope that you can use the keyword global in function definitions, to create and change the global names; similarly, in a nested function, you can use the keyword nonlocal to create and changes names in an enclosing scope.  

In [49]:
def outer():
    '''Prints the value of n'''
    n = 1
    
    def inner():
        nonlocal n
        n = 2
        print(n)
    
    inner()
    print(n)
    
    
outer()

2
2


**To summarize: name reference search at most 4 scopes, the local scope, then those of enclosing function, if there arent any, then global, then built-in.  This is know as LEGB rule.  Also remember that assigning names will only create or change local names. unless they are declared in global or nonlocal statements using keyword global or the keyword nonlocal respectively.  


## Nested Functions I

You've learned in the last video about nesting functions within functions. 
# One reason why you'd like to do this is to avoid writing out the same computations within functions repeatedly. 
There's nothing new about defining nested functions: you simply define it as you would a regular function with def and embed it inside another function!

In this exercise, inside a function three_shouts(), you will define a nested function inner() that concatenates a string object with !!!. three_shouts() then returns a tuple of three elements, each a string concatenated with !!! using inner(). Go for it!
Instructions
100 XP

    Complete the function header of the nested function with the function name inner() and 
# a single parameter word.
    Complete the return value: each element of the tuple should be a call to inner(), passing in the parameters from three_shouts() as arguments to each call.


In [50]:
# Define three_shouts
def three_shouts(word1, word2, word3):
    """Returns a tuple of strings
    concatenated with '!!!'."""

    # Define inner
    def inner(word):
        """Returns a string concatenated with '!!!'."""
        return word + '!!!'

    # Return a tuple of strings
    return (inner(word1), inner(word2), inner(word3))

# Call three_shouts() and print
print(three_shouts('a', 'b', 'c'))

('a!!!', 'b!!!', 'c!!!')


## Nested Functions II

Great job, you've just nested a function within another function. One other pretty cool reason for nesting functions is the idea of a closure. This means that the nested or inner function remembers the state of its enclosing scope when called. Thus, anything defined locally in the enclosing scope is available to the inner function even when the outer function has finished execution.

Let's move forward then! In this exercise, you will complete the definition of the inner function inner_echo() and then call echo() a couple of times, each with a different argument. Complete the exercise and see what the output will be!
Instructions
100 XP

    Complete the function header of the inner function with the function name inner_echo() and a single parameter word1.
    Complete the function echo() so that it returns inner_echo.
    We have called echo(), passing 2 as an argument, and assigned the resulting function to twice. Your job is to call echo(), passing 3 as an argument. Assign the resulting function to thrice.
    Hit Submit to call twice() and thrice() and print the results.


In [52]:
# Define echo
def echo(n):
    """Return the inner_echo function."""

    # Define inner_echo
    def inner_echo(word1):
        """Concatenate n copies of word1."""
        echo_word = word1 * n
        return echo_word

    # Return inner_echo
    return inner_echo

# Call echo: twice
twice = echo(2)

# Call echo: thrice
thrice = echo(3)



# Call twice() and thrice() then print
print(twice('hello'), thrice('hello'))

hellohello hellohellohello


## The keyword nonlocal and nested functions

Let's once again work further on your mastery of scope! In this exercise, you will use the keyword nonlocal within a nested function to alter the value of a variable defined in the enclosing scope.
Instructions
100 XP

    Assign to echo_word the string word, concatenated with itself.
    Use the keyword nonlocal to alter the value of echo_word in the enclosing scope.
    Alter echo_word to echo_word concatenated with '!!!'.
    Call the function echo_shout(), passing it a single argument 'hello'.


In [55]:
# Define echo_shout()
def echo_shout(word):
    """Change the value of a nonlocal variable"""
    
    # Concatenate word with itself: echo_word
    echo_word = word
    
    # Print echo_word
    print(echo_word)
    
    # Define inner function shout()
    def shout():
        """Alter a variable in the enclosing scope"""    
        # Use echo_word in nonlocal scope
        nonlocal echo_word
        
        # Change echo_word to echo_word concatenated with '!!!'
        echo_word = echo_word + '!!!'
    
    # Call function shout()
    shout()
    
    # Print echo_word
    print(echo_word)
    
    
    
# Call function echo_shout() with argument 'hello'
echo_shout('hello')

hello
hello!!!


## Default and flexible arguments




**Lets say you're writing a function that takes nultiple parameters and that there is often a common value for some of these parameters.  In this case, you would like to be able to call the function without explicitly specifying every parameters.  In other words yo would like some parameters to have default arguments that are used when it is not specified otherwise.  

In this video, you'll learn how to write function with default arguments, along with using flexible arguments, which allows you to pass any number of arguments to a function, as we'll soon see.  


First up, to define a function with a default argument value, in the function header, we follow the parameter of interest with an equal sign and the default argument value.  Notice that below function raise the first argument to the power of second of argument, and the default second argument value is 1.

In [57]:
def power(number, pow=2):
    '''Raise number to the power of pow'''
    new_value = number**pow
    return new_value


print(power(12))

print(power(12,3))

144
1728


# Let now look at flexible arguments, lets say that you want to write a function but arent sure how many arguments a user will want to pass it.  
For example, a function that takes floats or ints and adds them up, irrespective of how many there are.  Enter flexible arguments. 


In below example, we write the function that sums up all the arguments passed to it.  In the function definition, we use the parameter str followed by args: this then turns all the arguments passed to a function call into a tuple called args in the function body.  (*can be used on unpacking also on packing)  Then in the function body, to write our desired function, we initialize our sum_all to 0, loop over the tuple args and add each element of it successively to aum_all and then return it.  

We can now call our function add_all with any number of arguments to add them all up.  


In [61]:
def add_all(*args):
    '''Sum all values in *argstogether'''
    
    # Initialize sum
    sum_all = 0
    
    # Accumulate the sum
    for num in args:
        sum_all += num
        
    return sum_all


a = [1, 2, 3, 9.7, 23, 89, 6]
add_all(*a)

133.7


**You can also use a double star to pass an arbitrary number keyword arguments, also called kwargs, that is arguments preceded by identifiers.  We'll write such a function called print_all that prints out the identifiers and the parameters passed to them as you see here.  

Now, to write such function, we use the parameter kwargs preceded by double star.  This turns the identifier-keyword pairs into a dictionary withinn the function body.  Then in the function body all we need to do is to print out all the key-value pairs stored in the dictionary kwargs.  


Note that is NOT the name args and kwargs that are important when using flexible arguments, but rather that they are processed by a single and double star respectively.  


In [65]:
def print_all(**kwargs):
    '''Print out key-value pairs in **kwargs'''
    
    # Print out the key-value pairs
    for key, value in kwargs.items():
        print(key + ": " + value)



print_all(name='jj', employer='PP')

name: jj
employer: PP


## Functions with one default argument

In the previous chapter, you've learned to define functions with more than one parameter and then calling those functions by passing the required number of arguments. In the last video, Hugo built on this idea by showing you how to define functions with default arguments. You will practice that skill in this exercise by writing a function that uses a default argument and then calling the function a couple of times.
Instructions
100 XP

    Complete the function header with the function name shout_echo. It accepts an argument word1 and a default argument echo with default value 1, in that order.
    Use the * operator to concatenate echo copies of word1. Assign the result to echo_word.
    Call shout_echo() with just the string, "Hey". Assign the result to no_echo.
    Call shout_echo() with the string "Hey" and the value 5 for the default argument, echo. Assign the result to with_echo.


In [66]:
# Define shout_echo
def shout_echo(word1, echo=1):
    """Concatenate echo copies of word1 and three
     exclamation marks at the end of the string."""

    # Concatenate echo copies of word1 using *: echo_word
    echo_word = word1 * echo

    # Concatenate '!!!' to echo_word: shout_word
    shout_word = echo_word + '!!!'

    # Return shout_word
    return shout_word

# Call shout_echo() with "Hey": no_echo
no_echo = shout_echo('Hey')

# Call shout_echo() with "Hey" and echo=5: with_echo
with_echo = shout_echo('Hey', echo=5)

# Print no_echo and with_echo
print(no_echo)
print(with_echo)

Hey!!!
HeyHeyHeyHeyHey!!!


## Functions with multiple default arguments

You've now defined a function that uses a default argument - don't stop there just yet! You will now try your hand at defining a function with more than one default argument and then calling this function in various ways.

After defining the function, you will call it by supplying values to all the default arguments of the function. Additionally, you will call the function by not passing a value to one of the default arguments - see how that changes the output of your function!
Instructions
100 XP

    Complete the function header with the function name shout_echo. It accepts an argument word1, a default argument echo with default value 1 and a default argument intense with default value False, in that order.
    In the body of the if statement, make the string object echo_word upper case by applying the method .upper() on it.
    Call shout_echo() with the string, "Hey", the value 5 for echo and the value True for intense. Assign the result to with_big_echo.
    Call shout_echo() with the string "Hey" and the value True for intense. Assign the result to big_no_echo.


In [67]:
# Define shout_echo
def shout_echo(word1, echo=1, intense=False):
    """Concatenate echo copies of word1 and three
    exclamation marks at the end of the string."""

    # Concatenate echo copies of word1 using *: echo_word
    echo_word = word1 * echo

    # Make echo_word uppercase if intense is True
    if intense is True:
        # Make uppercase and concatenate '!!!': echo_word_new
        echo_word_new = echo_word.upper() + '!!!'
    else:
        # Concatenate '!!!' to echo_word: echo_word_new
        echo_word_new = echo_word + '!!!'

    # Return echo_word_new
    return echo_word_new

# Call shout_echo() with "Hey", echo=5 and intense=True: with_big_echo
with_big_echo = shout_echo('Hey', echo=5, intense=True)

# Call shout_echo() with "Hey" and intense=True: big_no_echo
big_no_echo = shout_echo('Hey', True)

# Print values
print(with_big_echo)
print(big_no_echo)

HEYHEYHEYHEYHEY!!!
Hey!!!


## Functions with variable-length arguments (*args)

Flexible arguments enable you to pass a variable number of arguments to a function. In this exercise, you will practice defining a function that accepts a variable number of string arguments.

The function you will define is gibberish() which can accept a variable number of string values. Its return value is a single string composed of all the string arguments concatenated together in the order they were passed to the function call. You will call the function with a single string argument and see how the output changes with another call using more than one string argument. Recall from the previous video that, within the function definition, args is a tuple.
Instructions
100 XP

    Complete the function header with the function name gibberish. It accepts a single flexible argument *args.
    Initialize a variable hodgepodge to an empty string.
    Return the variable hodgepodge at the end of the function body.
    Call gibberish() with the single string, "luke". Assign the result to one_word.
    Hit the Submit button to call gibberish() with multiple arguments and to print the value to the Shell.


In [72]:
# Define gibberish
def gibberish(*args):
    """Concatenate strings in *args together."""

    # Initialize an empty string: hodgepodge
    hodgepodge = str()

    # Concatenate the strings in args
    for word in args:
        hodgepodge = hodgepodge + word + ', '

    # Return hodgepodge
    return hodgepodge

# Call gibberish() with one string: one_word
one_word = gibberish('luke')

# Call gibberish() with five strings: many_words
many_words = gibberish("luke", "leia", "han", "obi", "darth")

# Print one_word and many_words
print(one_word)
print(many_words)

luke, 
luke, leia, han, obi, darth, 


## Functions with variable-length keyword arguments (**kwargs)

Let's push further on what you've learned about flexible arguments - you've used *args, you're now going to use **kwargs! What makes **kwargs different is that it allows you to pass a variable number of keyword arguments to functions. Recall from the previous video that, within the function definition, kwargs is a dictionary.

To understand this idea better, you're going to use **kwargs in this exercise to define a function that accepts a variable number of keyword arguments. The function simulates a simple status report system that prints out the status of a character in a movie.
Instructions
100 XP

    Complete the function header with the function name report_status. It accepts a single flexible argument **kwargs.
    Iterate over the key-value pairs of kwargs to print out the keys and values, separated by a colon ':'.
    In the first call to report_status(), pass the following keyword-value pairs: name="luke", affiliation="jedi" and status="missing".
    In the second call to report_status(), pass the following keyword-value pairs: name="anakin", affiliation="sith lord" and status="deceased".


In [79]:
# Define report_status
def report_status(**kwargs):
    """Print out the status of a movie character."""

    print("\n****BEGIN: REPORT****")

    # Iterate over the key-value pairs of kwargs
    for key, value in kwargs.items():
        # Print out the keys and values, separated by a colon ':'
        print(key + ": " + value)

    print("****END REPORT****\n")

# First call to report_status()
report_status(name='luke', affiliation='jedi', status='missing')

# Second call to report_status()
report_status(name='anakin', affiliation='sith lord', status='decreased')


****BEGIN: REPORT****
name: luke
affiliation: jedi
status: missing
****END REPORT****


****BEGIN: REPORT****
name: anakin
affiliation: sith lord
status: decreased
****END REPORT****



## Bringing it all together



**Its now time to get your hands dirty using you newly acquired skills by applying them to the wonderful world of DataScience: recall that, in the previous chapter, you wrote a function that would perform the following, it would take d DataFrame of Twitter data and return a dictionary containing langurage as keys and the number of times a tweet was written in a given language as values.  


# In the following exercises, you will write a function that processes a DataFrame and return a dictionary with counts of occurances in any column at all.  By default however, it will process a column calledlang.  

This generalizes the function that you wrote in previous exercises.  You will then generalize this further so that you can pass the function a DataFrame and any number of column names to perform the computation on an arbitrary number of columns.  



## Bringing it all together (1)

Recall the Bringing it all together exercise in the previous chapter where you did a simple Twitter analysis by developing a function that counts how many tweets are in certain languages. The output of your function was a dictionary that had the language as the keys and the counts of tweets in that language as the value.

In this exercise, we will generalize the Twitter language analysis that you did in the previous chapter. You will do that by including a default argument that takes a column name.

For your convenience, pandas has been imported as pd and the 'tweets.csv' file has been imported into the DataFrame tweets_df. Parts of the code from your previous work are also provided.
Instructions
100 XP

    Complete the function header by supplying the parameter for a DataFrame df and the parameter col_name with a default value of 'lang' for the DataFrame column name.
    Call count_entries() by passing the tweets_df DataFrame and the column name 'lang'. Assign the result to result1. Note that since 'lang' is the default value of the col_name parameter, you don't have to specify it here.
    Call count_entries() by passing the tweets_df DataFrame and the column name 'source'. Assign the result to result2.


In [100]:
# Define count_entries()
def count_entries(df, col_name='lang'):
    """Return a dictionary with counts of
    occurrences as value for each key."""

    # Initialize an empty dictionary: cols_count
    cols_count = {}

    # Extract column from DataFrame: col
    col = df[col_name]
    
    # Iterate over the column in DataFrame
    for entry in col:

        # If entry is in cols_count, add 1
        if entry in cols_count.keys():
            cols_count[entry] += 1

        # Else add the entry to cols_count, set the value to 1
        else:
            cols_count[entry] = 1

    # Return the cols_count dictionary
    return cols_count



import pandas as pd

tweets_df = pd.read_csv('tweets.csv')


# Call count_entries(): result1
result1 = count_entries(tweets_df)

# Call count_entries(): result2
result2 = count_entries(tweets_df, 'source')

# Print result1 and result2
print(result1)
print(result2)

{'en': 97, 'et': 1, 'und': 2}
{'<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>': 24, '<a href="http://www.facebook.com/twitter" rel="nofollow">Facebook</a>': 1, '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>': 26, '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>': 33, '<a href="http://www.twitter.com" rel="nofollow">Twitter for BlackBerry</a>': 2, '<a href="http://www.google.com/" rel="nofollow">Google</a>': 2, '<a href="http://twitter.com/#!/download/ipad" rel="nofollow">Twitter for iPad</a>': 6, '<a href="http://linkis.com" rel="nofollow">Linkis.com</a>': 2, '<a href="http://rutracker.org/forum/viewforum.php?f=93" rel="nofollow">newzlasz</a>': 2, '<a href="http://ifttt.com" rel="nofollow">IFTTT</a>': 1, '<a href="http://www.myplume.com/" rel="nofollow">Plume\xa0for\xa0Android</a>': 1}


## Bringing it all together (2)

Wow, you've just generalized your Twitter language analysis that you did in the previous chapter to include a default argument for the column name. You're now going to generalize this function one step further by allowing the user to pass it a flexible argument, that is, in this case, as many column names as the user would like!

Once again, for your convenience, pandas has been imported as pd and the 'tweets.csv' file has been imported into the DataFrame tweets_df. Parts of the code from your previous work are also provided.
Instructions
100 XP

#    Complete the function header by supplying the parameter for the dataframe df and the flexible argument *args.
    Complete the for loop within the function definition so that the loop occurs over the tuple args.
    Call count_entries() by passing the tweets_df DataFrame and the column name 'lang'. Assign the result to result1.
    Call count_entries() by passing the tweets_df DataFrame and the column names 'lang' and 'source'. Assign the result to result2.


In [103]:
# Define count_entries()
# *************************************************************************************************** #
def count_entries(df, *args):
    """Return a dictionary with counts of
    occurrences as value for each key."""
    
    #Initialize an empty dictionary: cols_count
    cols_count = {}
    
    # Iterate over column names in args
    for col_name in args:
    
        # Extract column from DataFrame: col
        col = df[col_name]
    
        # Iterate over the column in DataFrame
        for entry in col:
    
            # If entry is in cols_count, add 1
            if entry in cols_count.keys():
                cols_count[entry] += 1
    
            # Else add the entry to cols_count, set the value to 1
            else:
                cols_count[entry] = 1

    # Return the cols_count dictionary
    return cols_count




import pandas as pd

tweet_df = pd.read_csv('tweets.csv')
#print(tweet_df.head())


# Call count_entries(): result1
result1 = count_entries(tweet_df, 'lang')

# Call count_entries(): result2
result2 = count_entries(tweet_df, 'lang', 'source')

# Print result1 and result2
print(result1)
print(result2)

{'en': 97, 'et': 1, 'und': 2}
{'en': 97, 'et': 1, 'und': 2, '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>': 24, '<a href="http://www.facebook.com/twitter" rel="nofollow">Facebook</a>': 1, '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>': 26, '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>': 33, '<a href="http://www.twitter.com" rel="nofollow">Twitter for BlackBerry</a>': 2, '<a href="http://www.google.com/" rel="nofollow">Google</a>': 2, '<a href="http://twitter.com/#!/download/ipad" rel="nofollow">Twitter for iPad</a>': 6, '<a href="http://linkis.com" rel="nofollow">Linkis.com</a>': 2, '<a href="http://rutracker.org/forum/viewforum.php?f=93" rel="nofollow">newzlasz</a>': 2, '<a href="http://ifttt.com" rel="nofollow">IFTTT</a>': 1, '<a href="http://www.myplume.com/" rel="nofollow">Plume\xa0for\xa0Android</a>': 1}


## Lambda functions



**We've written our very own Python functions using the def keyword, function headers, docstrings and function bodies.  

# There is a quicker way to write functions on the fly, and these are called ambda functions because you use the keyword lambda.  



# ****************************************************************************************************
# Here we rewrite our function raise_to_power as a lambda function, to do so, after the keyword lambda, we specify the names of the arguments, then we use a colon followed by the expression that specifies what we wish the function to return.  


# ****************************************************************************************************
# **Lambda functions allow you to write functions in a quick and potentially dirty way so I wouldnt advice you to use them all the time but there are situations when they can come in very handy.  


# ****************************************************************************************************
For example, check out the map() function, which takes two arguments, a function and a sequence such as a list and applies the function over all elements of the sequence.  We can pass lambda functions to map without even naming them and in this case we refer to them as anonymous functions.  

# ****************************************************************************************************
In below example, we use map on a lambda function that squares all elements of a list and well store the result in square_all.  


In [80]:
raise_to_power = lambda x, y: x**y


raise_to_power(2, 3)

8

In [83]:
nums = [1, 2, 3, 4, 5, 6]


square_all = map(lambda x: x**2, nums)

[*square_all]

[1, 4, 9, 16, 25, 36]

## Pop quiz on lambda functions

In this exercise, you will practice writing a simple lambda function and calling this function. Recall what you know about lambda functions and answer the following questions:

    How would you write a lambda function add_bangs that adds three exclamation points '!!!' to the end of a string a?
    How would you call add_bangs with the argument 'hello'?

You may use the IPython shell to test your code.
Instructions
50 XP
Possible Answers

    The lambda function definition is: add_bangs = (a + '!!!'), and the function call is: add_bangs('hello').
#    The lambda function definition is: add_bangs = (lambda a: a + '!!!'), and the function call is: add_bangs('hello').
    The lambda function definition is: (lambda a: a + '!!!') = add_bangs, and the function call is: add_bangs('hello').


In [87]:
add_bangs = (lambda a: a + '!!!')
add_bangs2 = lambda a: a + '!!!'


print(add_bangs('Hello'))
add_bangs2('Bonjue')

Hello!!!


'Bonjue!!!'

## Writing a lambda function you already know

# Some function definitions are simple enough that they can be converted to a lambda function. 
By doing this, you write less lines of code, which is pretty awesome and will come in handy, especially when you're writing and maintaining big programs. In this exercise, you will use what you know about lambda functions to convert a function that does a simple task into a lambda function. Take a look at this function definition:

def echo_word(word1, echo):
    """Concatenate echo copies of word1."""
    words = word1 * echo
    return words

The function echo_word takes 2 parameters: a string value, word1 and an integer value, echo. It returns a string that is a concatenation of echo copies of word1. Your task is to convert this simple function into a lambda function.
Instructions
100 XP

    Define the lambda function echo_word using the variables word1 and echo. Replicate what the original function definition for echo_word() does above.
    Call echo_word() with the string argument 'hey' and the value 5, in that order. Assign the call to result.


In [88]:
# Define echo_word as a lambda function: echo_word
echo_word = (lambda word1, echo: word1*echo)

# Call echo_word: result
result = echo_word('hey', 5)

# Print result
print(result)

heyheyheyheyhey


## Map() and lambda functions

So far, you've used lambda functions to write short, simple functions as well as to redefine functions with simple functionality. The best use case for lambda functions, however, are for when you want these simple functionalities to be anonymously embedded within larger expressions. What that means is that the functionality is not stored in the environment, unlike a function defined with def. To understand this idea better, you will use a lambda function in the context of the map() function.

Recall from the video that map() applies a function over an object, such as a list. Here, you can use lambda functions to define the function that map() will use to process the object. For example:

nums = [2, 4, 6, 8, 10]

result = map(lambda a: a ** 2, nums)

You can see here that a lambda function, which raises a value a to the power of 2, is passed to map() alongside a list of numbers, nums. The map object that results from the call to map() is stored in result. You will now practice the use of lambda functions with map(). For this exercise, you will map the functionality of the add_bangs() function you defined in previous exercises over a list of strings.
Instructions
100 XP

    In the map() call, pass a lambda function that concatenates the string '!!!' to a string item; also pass the list of strings, spells. Assign the resulting map object to shout_spells.
    Convert shout_spells to a list and print out the list.


In [89]:
# Create a list of strings: spells
spells = ["protego", "accio", "expecto patronum", "legilimens"]

# Use map() to apply a lambda function over spells: shout_spells
shout_spells = map(lambda item: item + '!!!', spells)

# Convert shout_spells to a list: shout_spells_list
shout_spells_list = list(shout_spells)

# Print the result
print(shout_spells_list)

['protego!!!', 'accio!!!', 'expecto patronum!!!', 'legilimens!!!']


## Filter() and lambda functions

In the previous exercise, you used lambda functions to anonymously embed an operation within map(). You will practice this again in this exercise by using a lambda function with filter(), which may be new to you! 
# The function filter() offers a way to filter out elements from a list that don't satisfy certain criteria.

Your goal in this exercise is to use filter() to create, from an input list of strings, a new list that contains only strings that have more than 6 characters.
Instructions
100 XP

# ****************************************************************************************************
#    In the filter() call, pass a lambda function and the list of strings, fellowship. The lambda function should check if the number of characters in a string member is greater than 6; use the len() function to do this. Assign the resulting filter object to result.
    Convert result to a list and print out the list.


In [92]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'pippin', 'aragorn', 'boromir', 'legolas', 'gimli', 'gandalf']

# Use filter() to apply a lambda function over fellowship: result
# *************************************************************************************************** #
result = filter(lambda member: len(member)>6, fellowship)

# Convert result to a list: result_list
result_list = list(result)

# Print result_list
print(result_list)




print(help(filter))

['samwise', 'aragorn', 'boromir', 'legolas', 'gandalf']
Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.

None


## Reduce() and lambda functions

# ***************************************************************************************************
You're getting very good at using lambda functions! Here's one more function to add to your repertoire of skills. The reduce() function is useful for performing some computation on a list and, unlike map() and filter(), returns a single value as a result. To use reduce(), you must import it from the functools module.

Remember gibberish() from a few exercises back?

# Define gibberish
def gibberish(*args):
    """Concatenate strings in *args together."""
    hodgepodge = ''
    for word in args:
        hodgepodge += word
    return hodgepodge

gibberish() simply takes a list of strings as an argument and returns, as a single-value result, the concatenation of all of these strings. In this exercise, you will replicate this functionality by using reduce() and a lambda function that concatenates strings together.
Instructions
100 XP

    Import the reduce function from the functools module.
    In the reduce() call, pass a lambda function that takes two string arguments item1 and item2 and concatenates them; also pass the list of strings, stark. Assign the result to result. The first argument to reduce() should be the lambda function and the second argument is the list stark.


In [96]:
from functools import reduce

# Import reduce from functools

# Create a list of strings: stark
stark = ['robb', 'sansa', 'arya', 'brandon', 'rickon']

stark = [str(i+', ') for i in stark]

# Use reduce() to apply a lambda function over stark: result
result = reduce(lambda item1, item2: item1+item2, stark)

# Print the result
print(result)



print(help(reduce))

robb, sansa, arya, brandon, rickon, 
Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.

None


## Introduction to error handling



**When you use a function incorrectly, it should throw you an error.  



In [98]:
print(float(123))



print(float('123.876'))



print(float('Hw'))

123.0
123.876


ValueError: could not convert string to float: 'Hw'


When we write our own functions, we may wish to catch soecific problems and write specific error messages.  Lets check out this user-defined function that computes the square root of a number, it behaves as expected with integers.  What happens if we pass it a string such as 'bonjour sava'? 


Then it throws me an error corresponding to a line of code within the function definition.  This error says it was some sort of TypeError but the message may not be particularly useful to a user of our function, so we should endeavor to provide useful messages for the functions we write.  

# This is an example of an error caught during execution, commonly called exceptions.  The main way to catch such exceptions is the try-except clause, in which Python tries to run the code following try and if it can, all is well.  If it can not, due to an exception, it runs the code following except.  


Lets now rewrite our square root function but this time catch any exceptions raised.  

In [99]:
def sqrt(x):
    '''Returns the square root of a value'''
    return x**(0.5)



print(sqrt(12))


print(sqrt('bonjour sawa'))

3.4641016151377544


TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'float'

In [105]:
def sqrt(x):
    '''Returns the square root of an value'''
    try:
        return x**(0.5)
    except:
        print('keep arg as int or float please! ')
        
        
        
print(sqrt(12))

print(sqrt('bonjour sawa'))

3.4641016151377544
keep arg as int or float please! 
None


**We may also wish to only catch TypeError and let other error pass through, in which case we would use except TypeError as you can see below.  

There are many other types of exceptions that can be caught and you can get familiar with them by practice.  

In [115]:
def sqrt(x):
    '''Return the square root of an value'''
    
    try:
        return x**(0.5)
    except TypeError:
        print('lease keep arg x as int or float')
        
        
        
sqrt('bonjour')

x = [1, 2, 3, 4, 5, 6]
sqrt(x)

sqrt(-9993)

lease keep arg x as int or float
lease keep arg x as int or float


(6.121090488658852e-15+99.96499387285532j)

# **More often than not, instead of merely printing an error message, we'll want to actually raise an error by using he keyward raise.  

# ***************************************************************************************************
For example, our square root function does something we may not desire, when applied to negatuive numbers.  It actually returns a complex number which we may not want.  In fact, lets say that we dont wish our function to work for negative numbers.  Then using an if clause, we can raise a ValueError for cases in which he user passes the function a negative number.  


In [119]:
def sqrt(x):
    '''Return the square root of given number'''
    if isinstance(x, str):
        raise TypeError('Keep arg x not str please')
    
    elif x<0:
        raise ValueError('arg x must be positive value')
        
    try:
        return x**(0.5)
    except TypeError:
        print('Please keep arg x in int or float')
        
        
        
print(sqrt(32))

#print(sqrt('bonjour'))

print(sqrt(-9))

5.656854249492381


ValueError: arg x must be positive value

## Pop quiz about errors

In the video, Hugo talked about how errors happen when functions are supplied arguments that they are unable to work with. In this exercise, you will identify which function call raises an error and what type of error is raised.

Take a look at the following function calls to len():

len('There is a beast in every man and it stirs when you put a sword in his hand.')

len(['robb', 'sansa', 'arya', 'eddard', 'jon'])

len(525600)

len(('jaime', 'cersei', 'tywin', 'tyrion', 'joffrey'))

Which of the function calls raises an error and what type of error is raised?
Instructions
50 XP
Possible Answers

    The call len('There is a beast in every man and it stirs when you put a sword in his hand.') raises a TypeError.
    The call len(['robb', 'sansa', 'arya', 'eddard', 'jon']) raises an IndexError.
    The call len(525600) raises a TypeError.
    The call len(('jaime', 'cersei', 'tywin', 'tyrion', 'joffrey')) raises a NameError.
    

In [122]:
print(len('There is a beast in every man and it stirs when you put a sword in his hand.'))

print(len(['robb', 'sansa', 'arya', 'eddard', 'jon']))

print(len(525600))

print(len(('jaime', 'cersei', 'tywin', 'tyrion', 'joffrey')))

76
5


TypeError: object of type 'int' has no len()

In [125]:
print(dir(12345))


#print(help(12345))

['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__gt__', '__hash__', '__index__', '__init__', '__init_subclass__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'as_integer_ratio', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']


## Error handling with try-except

A good practice in writing your own functions is also anticipating the ways in which other people (or yourself, if you accidentally misuse your own function) might use the function you defined.

As in the previous exercise, you saw that the len() function is able to handle input arguments such as strings, lists, and tuples, but not int type ones and raises an appropriate error and error message when it encounters invalid input arguments. One way of doing this is through exception handling with the try-except block.

# In this exercise, you will define a function as well as use a try-except block for handling cases when incorrect input arguments are passed to the function.

Recall the shout_echo() function you defined in previous exercises; parts of the function definition are provided in the sample code. Your goal is to complete the exception handling code in the function definition and provide an appropriate error message when raising an error.
Instructions
100 XP

    Initialize the variables echo_word and shout_words to empty strings.
    Add the keywords try and except in the appropriate locations for the exception handling block.
    Use the * operator to concatenate echo copies of word1. Assign the result to echo_word.
    Concatenate the string '!!!' to echo_word. Assign the result to shout_words.


In [128]:
# Define shout_echo
def shout_echo(word1, echo=1):
    """Concatenate echo copies of word1 and three
    exclamation marks at the end of the string."""

    # Initialize empty strings: echo_word, shout_words
    echo_word = str()
    shout_words = str()
    
    
    
    if not isinstance(word1, str) or not isinstance(echo, int):
        raise TypeError('Please keep first arg word1 in str and second arg echo in int')
    
    
    
    # Add exception handling with try-except
    try:
        # Concatenate echo copies of word1 using *: echo_word
        echo_word = word1 * echo

        # Concatenate '!!!' to echo_word: shout_words
        shout_words = echo_word + '!!!'
    except ValueError:
        # Print error message
        print("word1 must be a string and echo must be an integer.")

    # Return shout_words
    return shout_words



# Call shout_echo
shout_echo("particle", echo="accelerator")

TypeError: Please keep first arg word1 in str and second arg echo in int

In [ ]:
# Solution:

# Define shout_echo
def shout_echo(word1, echo=1):
    """Concatenate echo copies of word1 and three
    exclamation marks at the end of the string."""

    # Initialize empty strings: echo_word, shout_words
    echo_word = ''
    shout_words = ''

    # Add exception handling with try-except
    try:
        # Concatenate echo copies of word1 using *: echo_word
        echo_word = word1 * echo

        # Concatenate '!!!' to echo_word: shout_words
        shout_words = echo_word + '!!!'
    except:
        # Print error message
        print("word1 must be a string and echo must be an integer.")

    # Return shout_words
    return shout_words

# Call shout_echo
shout_echo("particle", echo="accelerator")

## Error handling by raising an error

Another way to raise an error is by using raise. In this exercise, you will add a raise statement to the shout_echo() function you defined before to raise an error message when the value supplied by the user to the echo argument is less than 0.

The call to shout_echo() uses valid argument values. To test and see how the raise statement works, simply change the value for the echo argument to a negative value. Don't forget to change it back to valid values to move on to the next exercise!
Instructions
100 XP

    Complete the if statement by checking if the value of echo is less than 0.
    In the body of the if statement, add a raise statement that raises a ValueError with message 'echo must be greater than or equal to 0' when the value supplied by the user to echo is less than 0.


In [ ]:
# Define shout_echo
def shout_echo(word1, echo=1):
    """Concatenate echo copies of word1 and three
    exclamation marks at the end of the string."""

    # Raise an error with raise
    if echo < 0:
        raise ValueError('echo must be greater or equal to 0')

    # Concatenate echo copies of word1 using *: echo_word
    echo_word = word1 * echo

    # Concatenate '!!!' to echo_word: shout_word
    shout_word = echo_word + '!!!'

    # Return shout_word
    return shout_word



# Call shout_echo
shout_echo("particle", echo=5)

ValueError: echo must be greater or equal to 0

## Bringing it all together




**You're now tgoing to use your hard-earnt skills to write error messages into your DataFrame analyzer that you have been building up in previous chapters: lets say that a user of your function passes your function the name of a column that isnt a column in the DF that they passed it, you'll want them know.  



In [130]:
def sqrt(x):
    '''Return squrt value of given arg'''
    try:
        return x**(0.5)
    except:
        print('x must be int or float')
        
        
sqrt(12)

3.4641016151377544

In [134]:
def sqrt(x):
    '''Return sqrt root of given number'''
    
    if not isinstance(x, int):
        raise TypeError('please keep argx as a value')
    elif x < 0:
        raise ValueError('arg x must be non-negative')
        
    try:
        return x**(0.5)
    except TypeError:
        print('x must be int or float')
        
        
        
#sqrt(-9)
sqrt('bongour')

TypeError: please keep argx as a value

## Bringing it all together (1)

This is awesome! You have now learned how to write anonymous functions using lambda, how to pass lambda functions as arguments to other functions such as map(), filter(), and reduce(), as well as how to write errors and output custom error messages within your functions. You will now put together these learnings to good use by working with a Twitter dataset. Before practicing your new error handling skills,in this exercise, you will write a lambda function and use filter() to select retweets, that is, tweets that begin with the string 'RT'.

To help you accomplish this, the Twitter data has been imported into the DataFrame, tweets_df. Go for it!
Instructions
100 XP

    In the filter() call, pass a lambda function and the sequence of tweets as strings, tweets_df['text']. The lambda function should check if the first 2 characters in a tweet x are 'RT'. Assign the resulting filter object to result. To get the first 2 characters in a tweet x, use x[0:2]. To check equality, use a Boolean filter with ==.
    Convert result to a list and print out the list.


In [137]:
import pandas as pd

tweet_df = pd.read_csv('tweets.csv')



# Select retweets from the Twitter DataFrame: result
result = filter(lambda x: x[:2] == 'RT', tweet_df['text'])

# Create list from filter object result: res_list
res_list = list(result)

# Print all retweets in res_list
for tweet in res_list:
    print(tweet)

RT @bpolitics: .@krollbondrating's Christopher Whalen says Clinton is the weakest Dem candidate in 50 years https://t.co/pLk7rvoRSn https:/…
RT @HeidiAlpine: @dmartosko Cruz video found.....racing from the scene.... #cruzsexscandal https://t.co/zuAPZfQDk3
RT @AlanLohner: The anti-American D.C. elites despise Trump for his America-first foreign policy. Trump threatens their gravy train. https:…
RT @BIackPplTweets: Young Donald trump meets his neighbor  https://t.co/RFlu17Z1eE
RT @trumpresearch: @WaitingInBagdad @thehill Trump supporters have selective amnisia.
RT @HouseCracka: 29,000+ PEOPLE WATCHING TRUMP LIVE ON ONE STREAM!!!

https://t.co/7QCFz9ehNe
RT @urfavandtrump: RT for Brendon Urie
Fav for Donald Trump https://t.co/PZ5vS94lOg
RT @trapgrampa: This is how I see #Trump every time he speaks. https://t.co/fYSiHNS0nT
RT @trumpresearch: @WaitingInBagdad @thehill Trump supporters have selective amnisia.
RT @Pjw20161951: NO KIDDING: #SleazyDonald just attacked Scott Walker for NOT RAISI

## Bringing it all together (2)

Sometimes, we make mistakes when calling functions - even ones you made yourself. But don't fret! In this exercise, you will improve on your previous work with the count_entries() function in the last chapter by adding a try-except block to it. This will allow your function to provide a helpful message when the user calls your count_entries() function but provides a column name that isn't in the DataFrame.

Once again, for your convenience, pandas has been imported as pd and the 'tweets.csv' file has been imported into the DataFrame tweets_df. Parts of the code from your previous work are also provided.
Instructions
100 XP

    Add a try block so that when the function is called with the correct arguments, it processes the DataFrame and returns a dictionary of results.
    Add an except block so that when the function is called incorrectly, it displays the following error message: 'The DataFrame does not have a ' + col_name + ' column.'.


In [139]:
# Define count_entries()
def count_entries(df, col_name='lang'):
    """Return a dictionary with counts of
    occurrences as value for each key."""

    # Initialize an empty dictionary: cols_count
    cols_count = {}

    # Add try block
    try:
        # Extract column from DataFrame: col
        col = df[col_name]
        
        # Iterate over the column in dataframe
        for entry in col:
    
            # If entry is in cols_count, add 1
            if entry in cols_count.keys():
                cols_count[entry] += 1
            # Else add the entry to cols_count, set the value to 1
            else:
                cols_count[entry] = 1
    
        # Return the cols_count dictionary
        return cols_count

    # Add except block
    except:
        print('The DataFrame does not have a ' + col_name + ' column.')

# Call count_entries(): result1
result1 = count_entries(tweets_df, 'lang')

# Print result1
print(result1)



result2 = count_entries(tweets_df, 'hello')
print(result2)

{'en': 97, 'et': 1, 'und': 2}
The DataFrame does not have a hello column.
None


## Bringing it all together (3)

In the previous exercise, you built on your function count_entries() to add a try-except block. This was so that users would get helpful messages when calling your count_entries() function and providing a column name that isn't in the DataFrame. In this exercise, you'll instead raise a ValueError in the case that the user provides a column name that isn't in the DataFrame.

Once again, for your convenience, pandas has been imported as pd and the 'tweets.csv' file has been imported into the DataFrame tweets_df. Parts of the code from your previous work are also provided.
Instructions
100 XP

    If col_name is not a column in the DataFrame df, raise a ValueError 'The DataFrame does not have a ' + col_name + ' column.'.
    Call your new function count_entries() to analyze the 'lang' column of tweets_df. Store the result in result1.
    Print result1. This has been done for you, so hit 'Submit Answer' to check out the result. In the next exercise, you'll see that it raises the necessary ValueErrors.


In [141]:
# Define count_entries()
def count_entries(df, col_name='lang'):
    """Return a dictionary with counts of
    occurrences as value for each key."""
    
    # Raise a ValueError if col_name is NOT in DataFrame
    if col_name not in df.columns:
        raise ValueError('The DataFrame does not have a ' + col_name + ' column.')

    # Initialize an empty dictionary: cols_count
    cols_count = {}
    
    # Extract column from DataFrame: col
    col = df[col_name]
    
    # Iterate over the column in DataFrame
    for entry in col:

        # If entry is in cols_count, add 1
        if entry in cols_count.keys():
            cols_count[entry] += 1
            # Else add the entry to cols_count, set the value to 1
        else:
            cols_count[entry] = 1
        
        # Return the cols_count dictionary
    return cols_count

# Call count_entries(): result1
result1 = count_entries(tweets_df, 'lang')

# Print result1
print(result1)

{'en': 97, 'et': 1, 'und': 2}


## Bringing it all together: testing your error handling skills

You have just written error handling into your count_entries() function so that, when the user passes the function a column (as 2nd argument) NOT contained in the DataFrame (1st argument), a ValueError is thrown. You're now going to play with this function: it is loaded into pre-exercise code, as is the DataFrame tweets_df. Try calling count_entries(tweets_df, 'lang') to confirm that the function behaves as it should. Then call count_entries(tweets_df, 'lang1'): what is the last line of the output?
Instructions
50 XP
Possible Answers

    'ValueError: The DataFrame does not have the requested column.'
    'ValueError: The DataFrame does not have a lang1 column.'
    'TypeError: The DataFrame does not have the requested column.'
    

## Congratulations!



**